### PR TITLE
[Perf][Model] Optimize LTX video output transport

### PR DESCRIPTION
**Pull request by mglyn**
## Summary

Optimize the shared LTX-2/LTX-2.3/LTX-2.5 video output path:

- denormalize, clamp, scale, and round in the decoder output dtype;
- convert BCTHW directly to contiguous BTHWC `uint8` on GPU without a full-frame FP32 temporary;
- perform one D2H transfer before runner/IPC transport and use the existing API `uint8` fast path.

BF16/FP16 output bytes stay within one uint8 level of the legacy FP32 presentation path. Latents and generation are unchanged. Non-NumPy output types keep the existing `VideoProcessor` path.

This adapts the device-side uint8 transport approach from #6824 to LTX's BF16 decoded output and is part of #4985.

## Benchmark

NVIDIA H200; LTX-2.5 distilled two-stage tiled ConvVAE; `1920x1088x121`; 8+3 steps; eager CUDNN attention; fresh server per arm; one warmup plus three measured requests.

| Metric | FP32 temporary | Source dtype | Delta |
|---|---:|---:|---:|
| Incremental peak allocated | 3.532 GiB | 0.707 GiB | **-2.825 GiB** |
| Device output preparation, median | 9.616 ms | 5.776 ms | **1.66x faster** |
| Stage0, median | 17,898.1 ms | 17,924.4 ms | +0.15% |
| Server E2E, median | 18.543 s | 18.558 s | +0.08% |

The previous clean main-to-PR A/B measured the complete postprocess path at `1.284 s -> 9.566 ms` and E2E at `28.969 s -> 27.293 s`. The source-dtype follow-up removes the extra allocation while keeping that transport design and making the device conversion faster.

## Validation

- [x] All 10 official LTX similarity cases passed across LTX-2/LTX-2.3/LTX-2.5, distilled/full, one/two-stage, T2V/I2V.
- [x] Worst SSIM min `0.994486 >= 0.99`; worst PSNR `41.512 >= 40 dB`; worst audio relative L2 `0.015903 <= 0.10`; worst audio cosine `0.999874 >= 0.99`.
- [x] 122 focused LTX CPU tests and 4 CUDA transport-contract cases passed.
- [x] 4 E2E contract tests and applicable pre-commit hooks passed.

### CODE REVIEW

File: vllm_omni/diffusion/models/ltx2/ltx2_runtime.py [hunk 1]
```diff
@@ -84,6 +84,44 @@ def _run_ltx_vocoder(vocoder: nn.Module, generated_mel: torch.Tensor) -> torch.T
...
+        # Match VaeImageProcessor.denormalize in the source dtype. These
+        # in-place operations have the same dtype-rounding boundaries while
+        # avoiding another full-size decoded-video allocation.
+        video.mul_(0.5).add_(0.5).clamp_(0.0, 1.0)
+    else:
+        video.clamp_(0.0, 1.0)
+
+    # Diffusers converts to FP32 before NumPy output and the API server then
+    # scales and rounds in FP32. Preserve that order for byte-identical frames.
+    video = video.float()
```
Comment (david6666666) [inline vllm_omni/diffusion/models/ltx2/ltx2_runtime.py]: [P2] Avoid a full-frame FP32 temporary on GPU

At the 1920x1088x121 benchmark shape, `video.float()` is about 2.8 GiB and the uint8 result adds about 0.7 GiB while the decoded tensor is still live. That raises peak VRAM and scales with batch; could we convert in bounded chunks or move the FP32 quantization after a bounded D2H?

File: tests/diffusion/models/ltx2/test_ltx2_pipeline.py [hunk 2]
```diff
@@ -123,6 +128,98 @@ def _resolve_request_inputs_for_test(
...
+    ).reshape(2, 3, 4, 5, 7)
+    source = source.to(dtype)
+
+    legacy = processor.postprocess_video(source.clone(), output_type="np")
+    expected = np.clip(legacy, 0.0, 1.0)
+    expected *= 255.0
+    np.rint(expected, out=expected)
+    expected = expected.astype(np.uint8)
+
+    prepared = _prepare_ltx2_video_output(
```
Comment (david6666666) [inline tests/diffusion/models/ltx2/test_ltx2_pipeline.py:153]: [P2] Cover the GPU transport path

This test always passes CPU tensors, so it never exercises the CUDA conversion or the device-to-host handoff that this change is meant to optimize. Please add a GPU-marked case, or wire the reported CUDA parity check into CI, so a device-specific regression cannot pass here.

**Comment by mglyn** [comment]
## Follow-up: remove the full-frame FP32 temporary

### Cause

The first revision followed the MiniMax-H3 output-transfer pattern from #6824, including `video.detach().float()` before GPU quantization. That is harmless for MiniMax-H3 because its VAE already returns FP32 decoded frames. LTX's ConvVAE returns BF16 here, so copying that step mechanically upcast the full `[1, 3, 121, 1088, 1920]` video while the decoded tensor and final uint8 tensor were still live. The upcast alone is 2.825 GiB.

### New approach

Commit `5eb21dcea` keeps LTX denormalization, clamp, `* 255`, and rounding in the decoder output dtype, then converts BCTHW directly to contiguous BTHWC uint8 on the GPU. The existing single D2H/IPC path is unchanged, but no full-size FP32 CUDA tensor is materialized.

For BF16/FP16, presentation bytes can differ from the legacy FP32 presentation path by at most one uint8 level. This is after VAE decode and does not change latents or generation. The CPU and CUDA contract tests both enforce that bound.

### Results

Hardware: NVIDIA H200. Workload: LTX-2.5 distilled two-stage, tiled ConvVAE, `1920x1088x121`, 8+3 steps, eager CUDNN attention. Each E2E arm used a fresh server, one full-shape warmup, and three measured requests.

| Metric | Interim FP32 temporary | Source-dtype implementation | Delta |
|---|---:|---:|---:|
| Exact-shape incremental peak allocated | 3.532 GiB | 0.707 GiB | **-2.825 GiB** |
| Device output preparation, median | 9.616 ms | 5.776 ms | **-39.9%, 1.66x** |
| Engine postprocess, median | 412 ms | 200 ms | **-51.5%** |
| Stage0, median | 17,898.1 ms | 17,924.4 ms | +0.15% |
| Stage0 range | 2,188.8 ms | 64.2 ms | -97.1% |
| Server E2E, median | 18.543 s | 18.558 s | +0.08% |

The exact-shape allocation measurement isolates the conversion and proves that the full FP32 allocation is gone. The server's reserved-memory high-water mark is allocator-pool dependent; its warmup peak also dropped from 85,184 MiB to 82,290 MiB.

Self-review and validation on the updated commit:

- all 10 official LTX similarity cases passed across LTX-2, LTX-2.3, and LTX-2.5 (distilled/full, one/two-stage, T2V/I2V);
- worst video SSIM min was 0.994486 (threshold 0.99), worst PSNR was 41.512 dB (threshold 40 dB), worst audio relative L2 was 0.015903 (threshold 0.10), and worst audio cosine was 0.999874 (threshold 0.99);
- 122 focused LTX CPU tests and 4 CUDA transport-contract cases passed;
- 4 E2E request/reference contract tests and applicable pre-commit hooks passed.

This remains the LTX-specific part of #4985 and follows the device-side uint8 transport direction established by #6824, adapted to LTX's BF16 decoded output.

File: vllm_omni/diffusion/models/ltx2/ltx2_runtime.py (hunk 1 above)
Comment (mglyn) [reply vllm_omni/diffusion/models/ltx2/ltx2_runtime.py]: You're right — I should not have copied the MiniMax-H3 `video.float()` step mechanically. MiniMax-H3's decoded video is already FP32, while LTX's decoded video is BF16, so the LTX version created the full-frame FP32 temporary and raised peak memory.

I replaced it in `5eb21dcea` with source-dtype denormalization/scaling/rounding followed by a direct BF16/FP16-to-contiguous-BTHWC-uint8 conversion. At `1920x1088x121`, the isolated incremental allocated peak drops from 3.532 GiB to 0.707 GiB, removing 2.825 GiB, and median conversion time drops from 9.616 ms to 5.776 ms. The result is bounded to one uint8 level versus the legacy FP32 presentation path, with CPU and CUDA coverage; all 10 official LTX similarity cases pass.

Full explanation and results: https://github.com/vllm-project/vllm-omni/pull/7000#issuecomment-5529792470

File: tests/diffusion/models/ltx2/test_ltx2_pipeline.py (hunk 2 above)
Comment (mglyn) [reply tests/diffusion/models/ltx2/test_ltx2_pipeline.py:153]: Addressed in `5eb21dcea` with a separate GPU-marked test module: `tests/diffusion/models/ltx2/test_ltx2_output_cuda.py`.

The test now:

- creates the decoded-video input directly on CUDA;
- covers BF16 and FP16 with normalization enabled and disabled (4 cases);
- asserts that preparation returns contiguous BTHWC uint8 while still on CUDA;
- performs the D2H handoff and compares against the legacy presentation path with the documented one-level rounding bound.

I ran the module on NVIDIA H200: 4/4 cases passed. The complete official LTX similarity matrix also passed 10/10.

File: tests/diffusion/models/ltx2/test_ltx2_output_cuda.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        -1.2 if do_normalize else -0.2,
+        1.2,
+        2 * 3 * 4 * 5 * 7,
+        device="cuda",
+        dtype=torch.float32,
+    ).reshape(2, 3, 4, 5, 7)
+    source = source.to(dtype)
+
+    legacy = processor.postprocess_video(source.clone(), output_type="np")
+    expected = np.clip(legacy, 0.0, 1.0)
```
Comment (david6666666) [inline tests/diffusion/models/ltx2/test_ltx2_output_cuda.py]: [P2] Keep the `do_normalize=False` comparison honest

`VideoProcessor.postprocess_video(..., output_type="np")` does not clamp when `do_normalize=False`, but both new expectations call `np.clip`, matching the helper's clamp rather than the legacy result. With the `[-0.2, 1.2]` input, a regression can pass unnoticed. Could we compare against the unclipped legacy output (or keep this case within `[0, 1]`) and explicitly decide whether the helper should change that contract?

Comment (mglyn) [reply tests/diffusion/models/ltx2/test_ltx2_output_cuda.py]: Addressed in ce69066ee. The do_normalize=False parity cases now use valid [0, 1] decoder output and compare the unclipped VideoProcessor result. I also added a separate CUDA case that explicitly locks the intentional defensive clamp for out-of-range input. No production behavior changed. Targeted CPU tests (6/6), CUDA tests (7/7), and pre-commit pass.

**Comment by mglyn** [comment]
Here are the concrete LTX-2.5 official-reference similarity results for the source-dtype implementation:

| Case | SSIM mean | SSIM min | PSNR (dB) | Audio relative L2 | Audio cosine |
|---|---:|---:|---:|---:|---:|
| Distilled two-stage T2V | 0.997751 | 0.997448 | 50.646 | 0.015903 | 0.999874 |
| Distilled two-stage I2V | 0.998420 | 0.998365 | 54.264 | 0.003936 | 0.999994 |
| Full one-stage T2V | 0.997859 | 0.997193 | 45.476 | 0.004019 | 0.999993 |
| Full one-stage I2V | 0.997943 | 0.997903 | 52.669 | 0.006356 | 0.999981 |
| Full two-stage T2V | 0.995828 | 0.994486 | 41.512 | 0.003986 | 0.999993 |
| Full two-stage I2V | 0.997941 | 0.997899 | 52.248 | 0.007317 | 0.999975 |
| Required threshold | >= 0.99 | >= 0.99 | >= 40 | <= 0.10 | >= 0.99 |

All 6 LTX-2.5 cases passed. The test compares canonical in-memory video and audio arrays before MP4 encoding, so video encoder settings do not affect these measurements. The runs used NVIDIA H200 and the pinned official LTX-2.5 source revision `7954dcb0d986bdc36ef272564a9789ade07fcc65`.

**Review by david6666666 — APPROVED** [review]
LGTM
